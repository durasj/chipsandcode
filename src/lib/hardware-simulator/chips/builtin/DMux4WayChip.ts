import type Chip from '../Chip';
import type { ChipPin } from '../Chip';

/**
 * DMux4Way chip implemented natively
 */
class DMux4WayChip implements Chip {
  public readonly name = 'DMux4Way' as string;

  private in: boolean[];
  private sel: boolean[];
  private a: boolean[];
  private b: boolean[];
  private c: boolean[];
  private d: boolean[];

  constructor(private width = 1) {
    this.in = new Array(width).fill(false);
    this.sel = [false, false];
    this.a = new Array(width).fill(false);
    this.b = new Array(width).fill(false);
    this.c = new Array(width).fill(false);
    this.d = new Array(width).fill(false);
  }

  setInput(name: 'in' | 'sel', value: boolean[]) {
    this[name] = value;
  }

  getOutput(name: 'a' | 'b' | 'c' | 'd') {
    return this[name];
  }

  getPins() {
    // Constructs map on the fly
    // Pin inspection is less performance sensitive than actual execution

    return new Map([
      ['in', { type: 'output', width: this.width, state: this.in, connections: [] }],
      ['sel', { type: 'input', width: 2, state: this.sel, connections: [] }],
      ['a', { type: 'input', width: this.width, state: this.a, connections: [] }],
      ['b', { type: 'input', width: this.width, state: this.b, connections: [] }],
      ['c', { type: 'input', width: this.width, state: this.c, connections: [] }],
      ['d', { type: 'input', width: this.width, state: this.d, connections: [] }],
    ]) satisfies Map<string, ChipPin>;
  }

  public run() {
    const selPin = this.selToPin();
    for (const pin of ['a', 'b', 'c', 'd'] as const) {
      this[pin] = selPin === pin ? this.in : new Array(this.width).fill(false);
    }
  }

  private selToPin() {
    const sel = this.sel.map((v) => (v ? '1' : '0')).join('');

    switch (sel) {
      case '00':
        return 'a';
      case '01':
        return 'b';
      case '10':
        return 'c';
      case '11':
        return 'd';
      default:
        return 'a';
    }
  }
}

export default DMux4WayChip;
