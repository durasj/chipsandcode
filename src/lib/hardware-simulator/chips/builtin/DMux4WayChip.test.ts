import { describe, expect, it } from 'vitest';

import DMux4WayChip from './DMux4WayChip';

describe('DMux4Way Chip', () => {
  it('Provides interface info', async () => {
    const chip = new DMux4WayChip();

    expect(chip.name).toBe('DMux4Way');
    expect(chip.getPins()).toMatchSnapshot();
  });

  it('Has correct logic', async () => {
    const chip = new DMux4WayChip();

    const values = [
      [[false], [false, false], [false], [false], [false], [false]],
      [[true], [false, false], [true], [false], [false], [false]],
      [[false], [false, true], [false], [false], [false], [false]],
      [[true], [false, true], [false], [true], [false], [false]],
      [[false], [true, false], [false], [false], [false], [false]],
      [[true], [true, false], [false], [false], [true], [false]],
      [[false], [true, true], [false], [false], [false], [false]],
      [[true], [true, true], [false], [false], [false], [true]],
    ];

    values.forEach(([input, sel, a, b, c, d]) => {
      chip.setInput('in', input);
      chip.setInput('sel', sel);
      chip.run();

      expect(chip.getOutput('a')).toStrictEqual(a);
      expect(chip.getOutput('b')).toStrictEqual(b);
      expect(chip.getOutput('c')).toStrictEqual(c);
      expect(chip.getOutput('d')).toStrictEqual(d);
    });
  });
});
