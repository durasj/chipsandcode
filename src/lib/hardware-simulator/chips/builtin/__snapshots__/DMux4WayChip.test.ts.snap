// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DMux4Way Chip > Provides interface info 1`] = `
Map {
  "in" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "output",
    "width": 1,
  },
  "sel" => {
    "connections": [],
    "state": [
      false,
      false,
    ],
    "type": "input",
    "width": 2,
  },
  "a" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "b" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "c" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "d" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
}
`;
